File: benchmarks/reference/geos/adapter.py
"""GEOS-backed reference adapter through the pinned Shapely wheel."""

from __future__ import annotations

import json
import math
import sys
import time
from pathlib import Path

from shapely.geometry import GeometryCollection, Polygon
from shapely.ops import unary_union
from shapely.validation import make_valid


def close_ring(path):
    points = [tuple(point) for point in path]
    if points and points[0] != points[-1]:
        points.append(points[0])
    return points


def region(paths):
    geometries = []
    for path in paths:
        ring = close_ring(path)
        if len(ring) >= 4:
            geometries.append(make_valid(Polygon(ring)))
    return unary_union(geometries) if geometries else GeometryCollection()


def operation(name, subjects, clips):
    left = region(subjects)
    right = region(clips)
    if name == "intersection":
        return left.intersection(right)
    if name == "union":
        return left.union(right)
    if name == "difference":
        return left.difference(right)
    if name == "xor":
        return left.symmetric_difference(right)
    raise ValueError(f"unknown operation: {name}")


def rings(geometry):
    if geometry.geom_type == "Polygon":
        return [list(geometry.exterior.coords)] + [list(interior.coords) for interior in geometry.interiors]
    if hasattr(geometry, "geoms"):
        result = []
        for child in geometry.geoms:
            result.extend(rings(child))
        return result
    return []


def signature(geometry):
    raw_rings = [
        [[point[0], point[1]] for point in ring[:-1] if len(ring) > 1]
        for ring in rings(geometry)
    ]
    return json.dumps(raw_rings, separators=(",", ":"))


def main(workload_path):
    workload = json.loads(Path(workload_path).read_text(encoding="utf-8"))
    minimum_sample_time_ns = 2_000_000
    maximum_iterations_per_sample = 1 << 20
    print(json.dumps({
        "implementation": "geos-shapely",
        "samples": 25,
        "warmups": 3,
        "minimum_sample_time_ns": minimum_sample_time_ns,
    }))
    for test_case in workload["cases"]:
        for _ in range(3):
            operation(test_case["clip_type"], test_case["subjects"], test_case["clips"])
        result = GeometryCollection()
        iterations_per_sample = 1
        while True:
            started = time.perf_counter_ns()
            for _ in range(iterations_per_sample):
                result = operation(test_case["clip_type"], test_case["subjects"], test_case["clips"])
            elapsed = time.perf_counter_ns() - started
            if elapsed >= minimum_sample_time_ns or iterations_per_sample == maximum_iterations_per_sample:
                break
            iterations_per_sample = min(iterations_per_sample * 2, maximum_iterations_per_sample)
        timings = []
        for _ in range(25):
            started = time.perf_counter_ns()
            for _ in range(iterations_per_sample):
                result = operation(test_case["clip_type"], test_case["subjects"], test_case["clips"])
            timings.append((time.perf_counter_ns() - started) // iterations_per_sample)
        timings.sort()
        print(json.dumps({
            "id": test_case["id"],
            "status": "ok",
            "error": None,
            "median_ns": timings[len(timings) // 2],
            "p95_ns": timings[math.ceil(len(timings) * 0.95) - 1],
            "iterations_per_sample": iterations_per_sample,
            "ring_count": len(rings(result)),
            "signature": signature(result),
        }, separators=(",", ":")))


if __name__ == "__main__":
    main(sys.argv[1] if len(sys.argv) > 1 else "benchmarks/workloads.json")

File: benchmarks/reference/martinez/adapter.mjs
import fs from 'node:fs';
import process from 'node:process';
import * as polygonClipping from 'martinez-polygon-clipping';

const workloadPath = process.argv[2];
if (!workloadPath) {
  throw new Error('usage: node adapter.mjs <workloads.json>');
}

const workload = JSON.parse(fs.readFileSync(workloadPath, 'utf8'));
const operations = {
  intersection: (subjects, clips) => execute('intersection', subjects, clips),
  union: (subjects, clips) => execute('union', subjects, clips),
  difference: (subjects, clips) => execute('difference', subjects, clips),
  xor: (subjects, clips) => execute('xor', subjects, clips),
};

function execute(operation, subjects, clips) {
  const left = toGeometry(subjects);
  const right = toGeometry(clips);
  if (subjects.length === 0 || clips.length === 0) {
    if (operation === 'intersection') return [];
    if (operation === 'difference' && subjects.length === 0) return [];
    const nonEmpty = subjects.length === 0 ? right : left;
    return polygonClipping.union(nonEmpty, nonEmpty);
  }
  return polygonClipping[operation === 'difference' ? 'diff' : operation](left, right);
}

function toGeometry(paths) {
  return paths.map((path) => {
    const first = path[0];
    const last = path.at(-1);
    const ring = last && first[0] === last[0] && first[1] === last[1]
      ? path
      : path.concat([first]);
    return [ring];
  });
}

function signature(result) {
  const rings = (result ?? []).flatMap((polygon) => polygon).map((ring) => {
    const points = ring.slice();
    if (points.length > 1 && points[0][0] === points.at(-1)[0] && points[0][1] === points.at(-1)[1]) {
      points.pop();
    }
    return points;
  });
  return JSON.stringify(rings);
}

const minimumSampleTimeNs = 2_000_000n;
const maximumIterationsPerSample = 1 << 20;

process.stdout.write(`${JSON.stringify({
  implementation: 'martinez-polygon-clipping',
  revision: '0.8.1',
  samples: 25,
  warmups: 3,
  minimum_sample_time_ns: Number(minimumSampleTimeNs),
})}\n`);

for (const testCase of workload.cases) {
  const operation = operations[testCase.clip_type];
  for (let index = 0; index < 3; index += 1) operation(testCase.subjects, testCase.clips);
  let result = [];
  let iterationsPerSample = 1;
  while (true) {
    const started = process.hrtime.bigint();
    for (let index = 0; index < iterationsPerSample; index += 1) {
      result = operation(testCase.subjects, testCase.clips);
    }
    const elapsed = process.hrtime.bigint() - started;
    if (elapsed >= minimumSampleTimeNs || iterationsPerSample === maximumIterationsPerSample) break;
    iterationsPerSample = Math.min(iterationsPerSample * 2, maximumIterationsPerSample);
  }
  const timings = [];
  for (let index = 0; index < 25; index += 1) {
    const started = process.hrtime.bigint();
    for (let iteration = 0; iteration < iterationsPerSample; iteration += 1) {
      result = operation(testCase.subjects, testCase.clips);
    }
    timings.push(Number((process.hrtime.bigint() - started) / BigInt(iterationsPerSample)));
  }
  timings.sort((left, right) => left - right);
  const medianNs = timings[Math.floor(timings.length / 2)];
  const p95Ns = timings[Math.ceil(timings.length * 0.95) - 1];
  process.stdout.write(`${JSON.stringify({
    id: testCase.id,
    status: 'ok',
    error: null,
    median_ns: medianNs,
    p95_ns: p95Ns,
    iterations_per_sample: iterationsPerSample,
    signature: signature(result),
    ring_count: (result ?? []).flat().length,
  })}\n`);
}
